exaudfclient/base/python/python3/python_ext_dataframe.cc

```diff
@@ -25,32 +27,82 @@ extern "C" {
 #define PY_DATE (NPY_USERDEF+4)
 #define PY_NONETYPE (NPY_USERDEF+5)
 #define PY_BOOL (NPY_USERDEF+6)
+#define PY_FLOAT (NPY_USERDEF+7)
+#define PY_TIMESTAMP (NPY_USERDEF+8)
+
+std::map<std::string, int> pandasDTypeStrToNumpyCTypeMap {
 
-std::map<std::string, int> typeMap {
-    {"bool", NPY_BOOL},
     {"int", NPY_INT32},
     {"intc", NPY_INT32},
     {"intp", NPY_INT64},
     {"int8", NPY_INT8},
     {"int16", NPY_INT16},
     {"int32", NPY_INT32},
     {"int64", NPY_INT64},
+    {"int8[pyarrow]", NPY_OBJECT},
+    {"int16[pyarrow]", NPY_OBJECT},
+    {"int32[pyarrow]", NPY_OBJECT},
+    {"int64[pyarrow]", NPY_OBJECT},
+    
     {"uint8", NPY_UINT8},
     {"uint16", NPY_UINT16},
     {"uint32", NPY_UINT32},
     {"uint64", NPY_UINT64},
-    {"float", NPY_FLOAT64},
-    {"float16", NPY_FLOAT16},
+    {"uint8[pyarrow]", NPY_OBJECT},
+    {"uint16[pyarrow]", NPY_OBJECT},
+    {"uint32[pyarrow]", NPY_OBJECT},
+    {"uint64[pyarrow]", NPY_OBJECT},
+    
     {"float32", NPY_FLOAT32},
     {"float64", NPY_FLOAT64},
+    {"float", NPY_FLOAT32},
+    {"double", NPY_FLOAT64},
+    {"float32[pyarrow]", NPY_OBJECT},
+    {"float64[pyarrow]", NPY_OBJECT},
+    {"float[pyarrow]", NPY_OBJECT},
+    {"double[pyarrow]", NPY_OBJECT},
+     // We let numpy convert float16 to float (32 bit) and then use the C conversion from float to double, because a proper conversion from float16 to double in C is very complicated.
+    {"float16", NPY_FLOAT32},
+    {"halffloat", NPY_FLOAT32},
+    {"float16[pyarrow]", NPY_OBJECT},
+    {"halffloat[pyarrow]", NPY_OBJECT},
+
+    {"string[pyarrow]", NPY_OBJECT},
+    {"string[python]", NPY_OBJECT},
+    {"string", NPY_OBJECT},
+
+    {"bool[pyarrow]", NPY_OBJECT},
+    {"boolean", NPY_OBJECT},
+    {"bool", NPY_BOOL},
+
+    {"datetime64[ns]", NPY_DATETIME},
+    {"timestamp[ns, tz=UTC][pyarrow]", NPY_OBJECT},
+    
+    {"object", NPY_OBJECT},
+    
+    {"py_NAType", PY_NONETYPE},
+    {"py_NoneType", PY_NONETYPE},
+    {"py_bool", PY_BOOL},
     {"py_int", PY_INT},
+    {"py_float", PY_FLOAT},
     {"py_decimal.Decimal", PY_DECIMAL},
     {"py_str", PY_STR},
     {"py_datetime.date", PY_DATE},
-    {"datetime64[ns]", NPY_DATETIME},
-    {"object", NPY_OBJECT},
-    {"py_NoneType", PY_NONETYPE},
-    {"py_bool", PY_BOOL}
+    {"py_Timestamp", PY_TIMESTAMP}
+};
+
+std::map<int, std::string> numpyCTypeToNumpyDTypeStrMap {
+    {NPY_BOOL, "bool"},
+    {NPY_INT8, "int8"},
+    {NPY_INT16, "int16"},
+    {NPY_INT32, "int32"},
+    {NPY_INT64, "int64"},
+    {NPY_UINT8, "uint8"},
+    {NPY_UINT16, "uint16"},
+    {NPY_UINT32, "uint32"},
+    {NPY_UINT64, "uint64"},
+    {NPY_FLOAT32, "float32"},
```

Comment:
What abount NPY_FLOAT16?
https://numpy.org/doc/stable/reference/c-api/dtype.html#c.NPY_FLOAT16

Reply:
This is deliberate, because we don't support NPY_FLOAT16, in the previous mapping, we mapped float16 to float32. Which means it gets cast to float32 by numpy.